services/venice-controller/src/main/java/com/linkedin/venice/controller/kafka/protocol/serializer/NewSemanticUsageValidator.java

```diff
@@ -0,0 +1,253 @@
+ * </ol>
+ *
+ * If the object is not using new semantic, we return False. Otherwise, we return True along with the message
+ * in errorMessage. The traverser should stop traversing the object if the return value is True.
+ */
+public class NewSemanticUsageValidator {
+  private static final ThreadLocal<String> errorMessage = ThreadLocal.withInitial(() -> "");
+
+  /**
+   * This is the validator function that we will pass into SchemaDiffTraverser to validate the two nodes
+   * with different schemas.
+   */
+  private final BiFunction<Object, Pair<Schema.Field, Schema.Field>, Boolean> SEMANTIC_VALIDATOR =
+      (object, schemasPair) -> {
+        Schema.Field currentField = schemasPair.getFirst();
+        Schema.Field targetField = schemasPair.getSecond();
+
+        // If object is null, return false, no need to validate more
+        if (object == null) {
+          return false;
+        }
+
+        // Default behavior, check if the value is non-default
+        return isNonDefaultValue(object, currentField, targetField);
+      };
+
+  /**
+   * Get the semantic validator function.
+   */
+  public BiFunction<Object, Pair<Schema.Field, Schema.Field>, Boolean> getSemanticValidator() {
+    return SEMANTIC_VALIDATOR;
+  }
+
+  /**
+   * General method to check if the value is non-default for the given field.
+   * @param object the value to check
+   * @param currentField the field to check
+   * @param targetField the target field to check
+   * @return true if the value is non-default, false otherwise
+   */
+  public boolean isNonDefaultValue(Object object, Schema.Field currentField, Schema.Field targetField) {
+    switch (currentField.schema().getType()) {
+      case UNION:
+        return isNonDefaultValueUnion(object, currentField);
+      case ENUM:
+        return isNewEnumValue(object, currentField, targetField);
+      case FIXED:
+        if (currentField.schema().getFixedSize() != targetField.schema().getFixedSize()) {
+          return returnTrueAndLogError(
+              String.format("Field %s: has different fixed size", formatFieldName(currentField.name())));
+        }
+        return false;
+      default:
+        return isNonDefaultValueField(object, currentField);
+    }
+  }
+
+  /**
+   * Check if the value is a new enum value.
+   * @param object the value to check
+   * @param currentField the field to check with enum type and its current symbols
+   * @param targetField the field to check with enum type and its target symbols
+   * @return true if the value is a new enum value, false otherwise
+   */
+  public boolean isNewEnumValue(Object object, Schema.Field currentField, Schema.Field targetField) {
+    List<String> currentSymbols = currentField.schema().getEnumSymbols();
+    List<String> targetSymbols = targetField.schema().getEnumSymbols();
+    if (targetSymbols.stream().anyMatch(symbol -> !currentSymbols.contains(symbol) && symbol.equals(object))) {
```

Comment:
`symbol` is always String; what does "symbol.equals(object)" indicate here?

Reply:
`If symbol in targetSymbols. symbol not in currentSymbols and symbol equal to object
`means

If symbol in new_enum, symbol not in prev_enum, and symbol equal to object --> object is the new enum value.

I understand it is confusing, let me fix this.